src/blocks/agc.rs

```diff
@@ -157,10 +180,51 @@ where
     #[inline(always)]
     fn scale(&mut self, input: T) -> T {
         let output = input * T::from(self.gain).unwrap();
-        if !self.gain_locked {
-            self.gain +=
-                (self.reference_power - output.abs().to_f32().unwrap()) * self.adjustment_rate;
-            self.gain = self.gain.min(self.max_gain);
+        // if the input power is very low or very high compared to the reference power, we still want to reach the reference power quickly.
+        // Thus we should make the gain also dependent on the input_power in relation to the reference power.
+
+        // Gain is only exactly 1.0 when the AGC block is freshly initialized
+        // We then can set the gain to a suitable value to scale
+        // the input power to e.g. 99% of the reference power immediately.
+        if self.gain == 1.0 {
```

Comment:
I would avoid such tests. You could maybe add a flag, but I think such a mechanism could also lead to problems. Consider a sine with perfect power level, which also is zero (or close to zero) at the first sample. So this would make it worse. So I would consider to delete this initial adjustment.